feat: fma-generic-contracts (#140)
ethereum-optimism/design-docs: security/fma-generic-contracts.md

Reviewer: > As for creating a generic FMA for this, I think we should probably set the standard that mitigations and recovery paths are also documented within the FMA that references this one. Each upgrade isn't always the same, and each upgrade has to point towards actually fulfilling the mitigations in this doc + having planned recovery modes if they fall out of norm.

Agreed with the above comment from @clabby. Perhaps we update the FMA template to note this, and require providing links and/or completed action items for the mitigations and detections to make sure the work was actually done?

Author: I'll note it for now. Don't want to do too much more than port the notion doc to this repo in this PR so we can limit review cycles.

Reviewer: See https://github.com/ethereum-optimism/design-docs/pull/140#discussion_r1821103827

  - [Proxy Initialization Failure](#proxy-initialization-failure)
  - [New Implementation Overwrites (wrong) Slots](#new-implementation-overwrites-wrong-slots)
  - [Proxy Initialized to Wrong Values](#proxy-initialized-to-wrong-values)
  - [Can’t Upgrade Implementation](#cant-upgrade-implementation)
  - [Able to Reinitialize Implementation](#able-to-reinitialize-implementation)
  - [Backwards-incompatible ABIs](#backwards-incompatible-abis)
  - [Invalid `DisputeGameFactory.setImplementation` execution](#invalid-disputegamefactorysetimplementation-execution)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

### Generic items we need to take into account for any smart contract upgrade:

#### Proxy Initialization Failure

- **Description:**
  Unable to initialize the proxy due to a logic error.

- **Risk Assessment:**
  High severity since it allows arbitrary initialization.
  Low likelihood since the initialization is a minimal operation.

- **Mitigations:**
  Extensive unit and end-to-end testing for proxy initializations.
  Proxy initializations are verified through simulations prior to execution.

- **Detection:**

>   Call initialize(..args) or by simulating the effort.

Reviewer: ```suggestion
  Call `initialize(..args)` or by simulating the effort.
```

- **Recovery Path(s):**

>   Perform an upgrade to perform the initialization.

Reviewer: There is a variant where the initializer args we provided revert, but others do not, so an upgrade may not be necessary
```suggestion
  Call `initialize`, or upgrade the implementation to perform the initialization.
```

#### New Implementation Overwrites (wrong) Slots

- **Description:**
  An upgrade to a smart contract that inadvertently overwrites the wrong storage values in proxy.

- **Risk Assessment:**
  High severity due to arbitrary slot bounds.

>   Low likelihood since this is well looked out for.

Reviewer: To be honest we do not have sufficient checks in place for this, but it's something I am hoping to fix soon ™️

Author: Sounds good. Will leave as-is to make the diff with notion easier to navigate when you update.

Reviewer: To clarify, the work for better automated storage slot checks is not yet scheduled or prioritized, so I'd suggest adding a mitigation to manually compare the storage layout snapshots between the last release and this release

- **Mitigations:**
  Storage layouts are validated to only extend and not rewrite the slots of the previous version.

- **Detection:**

>   If this error were missed, it could be detected by looking at the storage values in the proxy or recognizing an inconsistency in variables stored in overwritten slots.

Reviewer: We have storage snapshots in the monorepo that require manual review on PRs, so it's easy to miss breaking changes because since it relies on careful PR review. More info in https://github.com/ethereum-optimism/superchain-ops/issues/339.

  Alternatively, we could detect it from monitoring or user reports that something seems off/incorrect.

- **Recovery Path(s):**

>   Deploy new implementation with a fixed storage layout and write new values to storage.

Reviewer: nit to clarify, because I initially read "fixed" as meaning "not dynamic" 😅
```suggestion
  Deploy new implementation with a corrected storage layout and write new values to storage.
```

  Requires an upgrade.

> #### Proxy Initialized to Wrong Values

Reviewer: Looks like some content has been taken out from this doc, such as the mention of `test_initialize_succeeds`?

- **Description:**
    The proxy is initialized with the wrong values.

- **Risk Assessment:**
  High severity since arbitrary values can be incorrectly used.
  Low likelihood since proxy initialization has a small surface area and is well verified/tested.

- **Mitigations:**
  Validate initialized values through extensive unit testing.
  Simulations prior to performing the upgrade, verifying the initialized values.

- **Detection:**
  A simulation can show an unintended initialized proxy value.

>   Monitoring or user reports.

Author: Agreed, but keeping this PR to a port of the notion doc to unblock the stage 1.4 FMA. A follow on PR should address this. I'd recommend putting up a ticket in this repository to track this suggestion and implementing the changes once this pr lands.

>   Monitoring or user reports.

Reviewer: Unresolving for the same reason as https://github.com/ethereum-optimism/design-docs/pull/140#discussion_r1821103827, since this is a valid detection for the stage 1.4 FMA ad that FMA depends on this PR :)

>   Monitoring or user reports.

Author: I'm very unfamiliar with the workings of "user reports" so I'm not able to answer these questions.

>   Monitoring or user reports.

Reviewer: There's no concrete thing here, @Ethnical is just saying that some user reports come through immunefi issues, others get raised through various other means. Something like this seems fine
```suggestion
  Monitoring or user reports, including Immunefi reports and other means of getting user feedback
```

- **Recovery Path(s):**
  Repair the upgrade path (if needed).
  Upgrade the smart contract.

#### Can’t Upgrade Implementation

- **Description:**
  Contracts are unable to be upgraded.

- **Risk Assessment:**
  High severity since it effectively bricks an arbitrary contract in the system.
  Low likelihood since contract upgrades are verified and follow industry-wide, standardized patterns.

- **Mitigations:**
  Validate through extensive unit testing.
  Perform simulations that demonstrate upgradability.

- **Detection:**
  Simulate a contract upgrade to check for failure.

- **Recovery Path(s):**
  Upgrade contracts and/or proxies.

>   Update `op-chain-ops` if needed.

Reviewer 2: I am not familiar with `op-chains-ops` but seems only working for L2 with genesis file. If this is the case we need to indicate this recovery path only works for the L2 and not for the L1.

Reviewer: Originally `op-upgrade` was part of `op-chain-ops`, but now `op-upgrade` is deprecated, so I don't think "Update `op-chain-ops`" is actually a recovery path here and this will need to be updated

> #### Able to Reinitialize Implementation

Reviewer: Should reference our `Initializeable*.t.sol` test files here

- **Description:**
  A third party is able to reinitialize the contract to arbitrary values.

- **Risk Assessment:**
  High severity since an arbitrary actor can perform an action with arbitrary values.
  Low likelihood since authentication is a critical, tested and validated property of the system.

- **Mitigations:**
  Validate through extensive unit testing.
  Fuzz testing with random values.

>   Invariant testing that authentication properties hold.

Reviewer: What invariant tests did you have in mind for reinitializing?

- **Detection:**

>   Simulate an arbitrary reinitialization.

Reviewer: The above "Alternatively, we could detect it from monitoring or user reports that something seems off/incorrect." applies here also

- **Recovery Path(s):**
  Perform an upgrade that prevents the reinitialization.

#### Backwards-incompatible ABIs

- **Description:**
  Functionality expected to hold across an upgrade or update breaks compatibility with prior systems.

- **Risk Assessment:**

>   Low severity since forward-facing systems replace backwards systems.

Reviewer: I'd say the actual reason this is low severity because it's a liveness failure, and not a safety failure

>   Low likelihood since ABI changes require intentional changes.

Reviewer: ABI changes can happen accidentally too, such as changing struct params, so we should rephrase this

- **Mitigations:**
  Validate ABI diffs.

- **Detection:**
  Breaking tools.

>   User reports.

Reviewer: A lot of the sentences here are very terse, it would be great if we can be a little more detailed to help simplify the work FMA authors will have to do per https://github.com/ethereum-optimism/design-docs/pull/140/files?diff=unified&w=0#r1817421106. For example, what tools might break, what kinds of users might report this?

Reviewer: The hardfork version of this doc (https://github.com/ethereum-optimism/design-docs/blob/main/security/fma-generic-hardfork.md) has some pre-written action items and some good details. We should expand this to a similar level of detail with the applicable template actions

Author: I've worded it more closely to the notion doc - I'm aiming to port that document over here with minimal changes so as to unblock the Stage 1.4 FMA.

Reviewer: Unresolving to align on FMA plan

The stage 1.4 FMA will need these changes (this comment + https://github.com/ethereum-optimism/design-docs/pull/140#discussion_r1817421106, potentially others in here) to accurately assess the failure modes, so merging this PR with minimal diff from notion doesn't unblock the FMA, as it just means the changes get done in a follow up PR instead

Whether we make these changes in this PR or a follow up doesn't matter to me, but I do want to make sure we're aligned on including these changes as part of the stage 1.4 FMA

Author: Makes sense to me. Let's keep this PR to be a minimal diff port of the notion doc, cleaning up where really needed. I'm keen on working on the Stage 1.4 FMA doc, but this doc goes well outside my expertise as I haven't touched smart contracts in well over a year.

Reviewer: Sounds good. Where are we tracking stage 1.4 FMA issues? Feel free to pull this comment out into an issue someplace, and if you need contracts help fleshing out some of these items let us know :)

Author: For sure - stage 1.4 FMA issues should definitely be in https://github.com/ethereum-optimism/design-docs/pull/132. If there are contract failure modes that you need to see in addition to those let me know - I've added everything to the best of my knowledge.

- **Recovery Path(s):**
  Update integration paths for backwards-compatibility.
  Some integrations (e.g. immutable contracts) may not be fixable.
  Perform an upgrade to fix breaking change.

> #### Invalid `DisputeGameFactory.setImplementation` execution

Reviewer 2: I was wondering why `Invalid DisputeGameFactory.setImplementation` is in this doc? 
As, all the previous bullet points are **generic upgrade practices** for smart contracts?
This one is specific to only FDG. Is there any particular reason?

Author: I may have misunderstood @mds1 [here](https://discord.com/channels/1244729134312198194/1298669460768423987/1299118978160922655) but I thought this was the intended outcome.

Reviewer 3: I don't think this is supposed to be in the doc, I think that from the other doc (specific fma for 1.4) he was saying to link to this doc in a section for generic upgrade failures

Reviewer: Sorry, I can see why you thought that from my comment, that should have said to link to it from the stage 1.4 FMA. I agree with @Ethnical and @tynes that this item is not specific to all contract upgrades so it shouldn't be here. We do already have this failure mode described in https://github.com/ethereum-optimism/design-docs/blob/main/security/fma-generic-hardfork.md#invalid-disputegamefactorysetimplementation-execution, so we can link to that from the specific stage 1.4 FMA

Author: Understood, will remove this section then, and like to `fma-generic-hardfork` in the stage 1.4 fma doc. Thanks all.

> - [./fma-generic-hardfork.md](./fma-generic-hardfork.md#Invalid-`DisputGameFactory.setImplementation`-execution)

Reviewer: this link did not take me to the header, I think it's:

```suggestion
This item is detailed in [./fma-generic-hardfork.md](./fma-generic-hardfork.md#invalid-disputegamefactorysetimplementation-execution)
```

Reviewer 2: In the doc: `./fma-generic-hardfork.md#Invalid-`DisputGameFactory.setImplementation`-execution`. 

_Detection: An un-executed safe transaction is easily detectable. In the case of a misconfigured game implementation, the op-dispute-mon will alert proofs-squad and security on any attempt to exploit this misconfiguration._

I would like to know `op-dispute-mon` what alerts will be triggered in the case of the  misconfigured implementation of the FDG is detected?

Reviewer: Unresolving this since I didn't see @Ethnical's question answered. I am unsure of the answer, but I do think it's out of scope for this "generic contract upgrades" FMA? Perhaps better asked in discord or slack?

Reviewer 2: Yes this is out of scope for the current fma for the generic contracts for sure!
We can resolve this one thx!

- **Detection:**
  A simulation can show an unintended initialized proxy value.
  Monitoring or user reports.

- **Recovery Path(s):**
  Repair the upgrade path (if needed).
  Upgrade the smart contract.

#### Can’t Upgrade Implementation

- **Description:**
  Contracts are unable to be upgraded.

- **Risk Assessment:**
  High severity since it effectively bricks an arbitrary contract in the system.
  Low likelihood since contract upgrades are verified and follow industry-wide, standardized patterns.

- **Mitigations:**
  Validate through extensive unit testing.
  Perform simulations that demonstrate upgradability.

- **Detection:**
  Simulate a contract upgrade to check for failure.

- **Recovery Path(s):**
  Upgrade contracts and/or proxies.
  Update `op-chain-ops` if needed.

#### Able to Reinitialize Implementation

- **Description:**
  A third party is able to reinitialize the contract to arbitrary values.

- **Risk Assessment:**
  High severity since an arbitrary actor can perform an action with arbitrary values.
  Low likelihood since authentication is a critical, tested and validated property of the system.

- **Mitigations:**
  Validate through extensive unit testing.

>   - See: [Initializable.t.sol](https://github.com/ethereum-optimism/optimism/blob/e6ef3a900c42c8722e72c2e2314027f85d12ced5/packages/contracts-bedrock/test/vendor/Initializable.t.sol).

Reviewer: There is also `InitializerOZv5.t.sol`, since this is the generic FMA we should mention both

Author: I've dug through ctb, but don't see any reference of `InitializerOZv5.t.sol`. Can we add this in a follow-on pr once this lands?

Reviewer: Yep, just make sure to track somewhere so we don't forget! 

It's here: https://github.com/ethereum-optimism/optimism/blob/4b1c12ad4312fe6fe28d8297a26cccf053356704/packages/contracts-bedrock/test/vendor/InitializableOZv5.t.sol

Author: I've added it, thank you for linking!

  Fuzz testing with random values.
  Invariant testing that authentication properties hold.

- **Detection:**
  Simulate an arbitrary reinitialization.
  Alternatively, we could detect it from monitoring or user reports that something seems off/incorrect.

- **Recovery Path(s):**
  Perform an upgrade that prevents the reinitialization.

#### Backwards-incompatible ABIs

- **Description:**
  Functionality expected to hold across an upgrade or update breaks compatibility with prior systems.

- **Risk Assessment:**

>   Low severity since liveliness is not impacted.

Reviewer: ```suggestion
  Low severity since only liveness is impacted.
```

>   Low likelihood since ABI changes are validated.

Reviewer: ```suggestion
  Low likelihood since ABI changes are validated as part of our snapshot process during PR reviews.
```
